cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/WrappedSQLEngineCollection.java

```diff
@@ -166,7 +178,57 @@ public Runnable createMultiStoreTask(PhaseSpec phaseSpec,
                                        Set<String> group,
                                        Set<String> sinks,
                                        Map<String, StageStatisticsCollector> collectors) {
-    return unwrap().createMultiStoreTask(phaseSpec, group, sinks, collectors);
+    return new Runnable() {
+      @Override
+      public void run() {
+        // Copy all sinks and groups into a synchronized set
+        Set<String> remainingGroupStages = Collections.synchronizedSet(new HashSet<>(group));
+        Set<String> remainingSinks = Collections.synchronizedSet(new HashSet<>(sinks));
+
+        // Create list to store all tasks.
+        List<Future<Void>> directStoreFutures = new ArrayList<>();
+
+        // Try to run the direct store task on all sink stages.
+        for (String sinkName : sinks) {
+          StageSpec stageSpec = phaseSpec.getPhase().getStage(sinkName);
+
+          // Check if we are able to write this output directly
+          if (stageSpec != null) {
+            // Create an async task that is used to wait for the direct store task to complete.
+            // We submit these in parallel to prevent blocking for each store task to complete in sequence.
+            Runnable task = () -> {
+              if (wrapped.tryStoreDirect(stageSpec)) {
+                remainingGroupStages.remove(sinkName);
+                remainingSinks.remove(sinkName);
+              }
+            };
+            Future<Void> future = CompletableFuture.runAsync(task);
+            directStoreFutures.add(future);
+          }
+        }
+
+        // Wait for all the direct store tasks for this group, if any.
+        for (Future<Void> future : directStoreFutures) {
+          try {
+            future.get();
+          } catch (InterruptedException e) {
+            LOG.error("Interrupted exception when waiting for Direct Store task to complete", e);
+            throw Throwables.propagate(e);
+          } catch (ExecutionException e) {
+            // We don't propagate this exception as the regular sink workflow can continue.
+            LOG.error("Execution exception when waiting for Direct Store task to complete", e);
```

Comment:
If we don't stop the execution and actually handle it, could you make it a warning?

Reply:
Good point. I'll make this a warning instead.

Reply:
Updated